src/key_io.cpp

```diff
@@ -78,82 +78,128 @@ class DestinationEncoder
             return ScriptHash(hash);
         }
-
         // If the prefix of data matches either the script or pubkey prefix, the length must have been wrong
-        if ((data.size() >= script_prefix.size() &&
-                std::equal(script_prefix.begin(), script_prefix.end(), data.begin())) ||
-            (data.size() >= pubkey_prefix.size() &&
-                std::equal(pubkey_prefix.begin(), pubkey_prefix.end(), data.begin()))) {
+        if ((base58Data.size() >= script_prefix.size() &&
+                std::equal(script_prefix.begin(), script_prefix.end(), base58Data.begin())) ||
+            (base58Data.size() >= pubkey_prefix.size() &&
+                std::equal(pubkey_prefix.begin(), pubkey_prefix.end(), base58Data.begin()))) {
             error_str = "Invalid length for Base58 address (P2PKH or P2SH)";
         } else {
-            error_str = "Invalid or unsupported Base58-encoded address.";
+            std::string chainPrefixes = params.GetChainTypeString() == "main" ? "1 or 3" : "m, n, or 2";
+            error_str = strprintf("Invalid or unsupported Base58 %s address. Expected prefix %s", networkLabel, chainPrefixes);
         }
         return CNoDestination();
     } else if (!is_bech32) {
-        // Try Base58 decoding without the checksum, using a much larger max length
-        if (!DecodeBase58(str, data, 100)) {
-            error_str = "Invalid or unsupported Segwit (Bech32) or Base58 encoding.";
-        } else {
-            error_str = "Invalid checksum or length of Base58 address (P2PKH or P2SH)";
+        if (!is_base58) {
+            error_str = is_validBech32Chars ? "Bech32(m) address decoded with error: " + bech32Error : "Address is not valid Base58 or Bech32";
+            if (error_locations) *error_locations = std::move(bech32ErrorLoc);
+        }
+        else {
+            error_str = is_validBech32Chars
+                ? "Invalid address encoded as Base58 and Bech32(m) provided"
+                : "Invalid checksum or length of Base58 address (P2PKH or P2SH)";
+            if (is_validBech32Chars && error_locations) {
+                *error_locations = std::move(bech32ErrorLoc);
+            }
         }
         return CNoDestination();
     }
 
-    data.clear();
-    const auto dec = bech32::Decode(str);
-    if ((dec.encoding == bech32::Encoding::BECH32 || dec.encoding == bech32::Encoding::BECH32M) && dec.data.size() > 0) {
+    if ((bech32Encoding == bech32::Encoding::BECH32 || bech32Encoding == bech32::Encoding::BECH32M) && bech32Chars.size() > 0) {
         // Bech32 decoding
-        if (dec.hrp != params.Bech32HRP()) {
-            error_str = strprintf("Invalid or unsupported prefix for Segwit (Bech32) address (expected %s, got %s).", params.Bech32HRP(), dec.hrp);
+        if (bech32Hrp != params.Bech32HRP()) {
+            error_str = strprintf("Invalid chain prefix for %s. Expected %s got %s", networkLabel, params.Bech32HRP(), bech32Hrp);
```

Comment:
The prefix does not necessarily indicate a chain. The current error is more correct.

Reply:
This has been resolved by rewording the error message since you are correct addresses have prefixes not chains. 

```Invalid prefix for regtest address (expected bcrt, got bc)```

Reply:
Could just as well still be unsupported,  not invalid.

Reply:
I have changed this to read ```Invalid or unsupported prefix```... testing my own knowledge; The reason you can have an unsupported but still technically valid address would be in the event of a softfork to a new SegWit version? e.x. _a segwit client interpreting a taproot address would say "invalid"_ but that is not completely true because the address is valid. Just not supported by the client?

This has been resolved in b66e974ac8f21971602051c08df830d14cf3df7e

Reply:
Doesn't even have to be a soft fork. E.g. a silent payment address (#27827) has a prefix we don't support (yet), but it's not a soft fork. It's also not a different network.